### DISCUSSION TITLE
ssreflect views

### DISCUSSION THREAD

**Discussion by jonsterling** [jonsterling/coq-domains#7] [Q&A]
@clayrat Thanks for your help with this! I had a question about something that I wasn't able to understand from reading the ssreflect documentation. How exactly do those views in the `apply` tactic work? I mean like `apply/eq_sig/proof_irrelevance`.

Does this expand into some sequence of tactics not involving the slash? Or is it more complicated? Anyway the reason I ask is that I have trouble predicting when a sequence of views like this will do what I need, but it is very cool and I'd like to know how to use it.

Thanks!

**Comment by clayrat** [comment]
I think [this part](https://coq.inria.fr/refman/proof-engine/ssreflect-proof-language.html#general-mechanism-for-interpreting-goals-and-assumptions) describes it :

> A goal interpretation view tactic of the form:
>    `apply/term`
> applied to a goal `top` is interpreted in the following way:
> * If the type of `term` is not an instance of the `reflect` predicate, nor an equivalence, then the term `term` is applied to the current goal `top`, possibly inserting implicit arguments.

So `apply/foo/bar/baz` is basically a way of compressing a sequence of `apply: foo; apply: bar; apply: baz`. 

This gets a bit non-intuitive however, when a previous `apply` generates several goals, because as the docs say, it'll keep applying to the **top** goal (i.e., the last one). So if `foo` generates three goals, then `apply/foo/bar/baz/fnord` is equivalent to `apply: foo; [apply: fnord | apply: baz | apply: bar]` - note the reversal.

**Reply by jonsterling** [reply]
Thanks, but i’m still a little confused. I understood from the docs that subsequent views get applied to the principle goal, but i thought that should be the first goal rather than the last goal. Can you help me understand where i’m going wrong?

**Reply by jonsterling** [reply]
(Btw while I'm still not getting the total big picture, I am having fun changing my scripts to use this very compact notation, lol. so thanks!)

**Reply by clayrat** [reply]
To be honest, I'm not sure myself why is this a default behaviour ;) Let me ask on Coq zulip.

**Reply by jonsterling** [reply]
From a practical point of view, I'm finding it a very useful idiom because often a lemma unleashes some auxiliary goal that I just want to wipe out right away. So it's pretty nice... But I think that there is probably some big 🧠  SSREFLECT GEOMETRY 🧠  that is guiding this design too.
